### PR TITLE
feat(jest): add types for implementing custom interfaces

### PR DESCRIPTION
**Pull request by ikatyang**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <https://github.com/facebook/jest/tree/dd6c5c4/types>
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Review by orta — APPROVED** [review]
That's a lot of whitespace changes, dang. A good chunk of them are from me :P

I'm glad you tackled the internal objects, those are things that I've had to manually make interfaces for subsets in DangerJS. 👍

**Comment by RyanCavanaugh** [comment]
@ikatyang Please address the merge conflict.

File: types/jest/index.d.ts [hunk 1]
```diff
@@ -466,3 +702,369 @@ declare namespace jasmine {
         [n: number]: T;
     }
 }
+
+declare namespace jest {
+    // types for implementing custom interfaces, from https://github.com/facebook/jest/tree/dd6c5c4/types
+
+    // https://facebook.github.io/jest/docs/en/configuration.html#transform-object-string-string
+    // const transformer: Transformer;
```
Comment (jwbay) [inline types/jest/index.d.ts]: Are these commented out declarations helpful? Seems like something that would belong in the test file

File: types/jest/index.d.ts [hunk 2]
```diff
@@ -466,3 +702,369 @@ declare namespace jasmine {
...
+    type TestResultsProcessor = (testResult: AggregatedResult) => AggregatedResult;
+
+    type HasteResolver = any; // import HasteResolver from 'jest-resolve';
+    type ModuleMocker = any; // import {ModuleMocker} from 'jest-mock';
+    type ModuleMap = any; // import {ModuleMap} from 'jest-haste-map';
+    type HasteFS = any; // import {FS as HasteFS} from 'jest-haste-map';
+    type Runtime = any; // import Runtime from 'jest-runtime';
+    type Script = any; // import {Script} from 'vm';
+
+    type Obj = any; // `object` needs TS v2.2
```
Comment (jwbay) [inline types/jest/index.d.ts]: 2.1 -> 2.2 (or 2.3) is likely a fine change to make since TS is on 2.4 now.

File: types/jest/index.d.ts [hunk 3]
```diff
@@ -466,3 +702,369 @@ declare namespace jasmine {
...
+        runAllTicks(): void;
+        runAllTimers(): void;
+        runTimersToTime(msToRun: number): void;
+        runOnlyPendingTimers(): void;
+        runWithRealTimers(callback: any): void;
+        useFakeTimers(): void;
+        useRealTimers(): void;
+    }
+
+    interface $JestEnvironment {
```
Comment (jwbay) [inline types/jest/index.d.ts]: Why not `interface Environment` and kill the type alias?

Comment (ikatyang) [inline types/jest/index.d.ts]: It's the name from jest repo, I think we should not change it for better comparing in the future.

File: types/jest/index.d.ts (hunk 2 above)
Comment (ikatyang) [inline types/jest/index.d.ts]: done

**Comment by ikatyang** [comment]
@RyanCavanaugh done

**Review by jwbay — APPROVED** [review]

**Comment by NoHomey** [comment]
LGTM :+1:

**Comment by andy-ms** [comment]
Thanks!

**Pull request merged**